infra/base-images/base-builder/bisect_clang.py

```diff
@@ -0,0 +1,234 @@
+    assert label in ('good', 'bad'), label
+    self.do_command(['checkout', commit])
+    build_clang()
+    retcode, _, _ = execute(test_command, shell=True, expect_zero=False)
+    if label == 'good' and retcode != 0:
+      raise BisectException(retcode, commit)
+    if label == 'bad' and retcode == 0:
+      raise BisectException(
+          'Test Command returns %d on first bad commit %s' % (retcode, commit))
+
+    self.do_bisect_command(label)
+
+
+class BisectException(Exception):
+  pass
+
+
+def get_clang_build_env():
+  """Get an environment for building Clang."""
+  env = os.environ.copy()
+  for variable in ['CXXFLAGS', 'CFLAGS']:
+    if variable in env:
+      del env[variable]
+  return env
+
+
+def install_clang_build_deps():
+  """Instal dependencies necessary to build clang."""
+  execute([
+      'apt-get', 'install', '-y', 'build-essential', 'make', 'cmake',
+      'ninja-build', 'git', 'subversion', 'g++-multilib'
+  ])
+
+
+def clone_with_retries(repo, local_path, num_retries=10):
+  """Clone |repo| to |local_path| if it doesn't exist already. Try up to
+  |num_retries| times. Return False if unable to checkout."""
+  if os.path.isdir(local_path):
+    return
+  for _ in range(num_retries):
+    if os.path.isdir(local_path):
+      shutil.rmtree(local_path)
+    retcode, _, _ = execute(
+        ['git', 'clone', repo, local_path], expect_zero=False)
+    if retcode == 0:
+      return
+  raise Exception('Could not checkout %s.' % repo)
+
+
+def get_target_to_build():
+  """Get target architecture we want clang to target when we build it."""
+  _, arch, _ = execute(['uname', '-m'])
+  if 'x86_64' in arch:
+    return 'X86'
+  if 'aarch64' in arch:
+    return 'AArch64'
+  raise Exception('Unsupported target: %s.' % arch)
+
+
+def prepare_build(llvm_project_path):
+  """Prepare to build clang."""
+  llvm_build_dir = os.path.join(os.getenv('WORK'), 'llvm-build')
+  if not os.path.exists(llvm_build_dir):
+    os.mkdir(llvm_build_dir)
+  previous_dir = os.getcwd()
+  os.chdir(llvm_build_dir)
+  try:
+    execute(
+        [
+            'cmake', '-G', 'Ninja', '-DLIBCXX_ENABLE_SHARED=OFF',
+            '-DLIBCXX_ENABLE_STATIC_ABI_LIBRARY=ON',
+            '-DLIBCXXABI_ENABLE_SHARED=OFF', '-DCMAKE_BUILD_TYPE=Release',
+            '-DLLVM_ENABLE_PROJECTS=libcxx;libcxxabi;compiler-rt;clang',
+            '-DLLVM_TARGETS_TO_BUILD=' + get_target_to_build(),
+            os.path.join(llvm_project_path, 'llvm')
+        ],
+        env=get_clang_build_env())
+  finally:
+    os.chdir(previous_dir)
+
+  return llvm_build_dir
+
+
+def build_clang():
+  """Checkout, build and install Clang."""
+  # TODO(metzman): Merge Python checkout and build code with
+  # checkout_build_install_llvm.sh.
+  install_clang_build_deps()
+  # llvm_checkout_script = os.path.join(os.getenv('SRC'), 'checkout_llvm.sh')
+  llvm_project_path = os.path.join(os.getenv('SRC'), 'llvm-project')
+  clone_with_retries('https://github.com/llvm/llvm-project.git',
```

Comment:
nit: checkout once at the very start, rather than "lazily" checking out and bailing once it's already checked out